adam/learner/contrastive_learner.py

```diff
@@ -181,24 +190,22 @@ def learn_from(self, matching: LanguagePerceptionSemanticContrast) -> None:
         for concept in [concept1, concept2]:
             self._propose_updated_hypothesis_to_apprentice(concept)
 
-    def _match_concept_pattern_to_graph(
+    def _match_concept_pattern_to_multiple_graphs(
         self, concept: Concept, graph: PerceptionGraph
-    ) -> Optional[PerceptionGraphPatternMatch]:
-        # Arbitrarily take the first match.
-        #
-        # Because there may be more than one match, this results in non-deterministic behavior in
-        # general. We would like to do something smarter. See the GitHub issue:
-        # TODO issue
-        return (
-            self._get_apprentice_pattern(concept)
-            .matcher(
+    ) -> Optional[
+        Mapping[PerceptionGraphTemplate, Optional[PerceptionGraphPatternMatch]]
+    ]:
+        return {
+            apprentice_concept: apprentice_concept.graph_pattern.matcher(
```

Comment:
Ah, missed this earlier -- please retain the previous comment about arbitrarily taking the first match and replace `TODO issue` with a link to #1140. That comment is about an issue unrelated to #1138, one that is more to do with how graph matching works.

Reply:
On this, I think the issue changes a bit when handling multiple hypotheses. If there is more than one hypothesis, then the first `top_n` are parsed. These `top_n` still face the same arbitrary issues as prior (taking `_concept_to_hypotheses[:top_n]`), instead of just `_concept_to_hypotheses[0]`)

Reply:
The problem isn't with which or how many arbitrary hypotheses we take. The problem is that the matching process itself is not deterministic. Specifically, we don't have any guarantees that two seemingly identical hypothesis won't be matched differently to the same perception, or the same hypothesis won't be matched differently to two (nearly) identical perceptions.

Reply:
Feel free to update the comment to make this more clear.

Reply:
Ah, got it.